### DISCUSSION TITLE
Is there a sanctioned way to list sibling states for a given state to "know" about in a less... interesting way?

### DISCUSSION THREAD

**Discussion by Morklympious** [statelyai/xstate#3211] [General] [answered]
Hey y'all! I'm trying to conceptualize how a state might be "aware" of its siblings in a tabbed page sort of sense. My problem: I'm dealing with a component that has a nav whose buttons will navigate to other sibling states. 

Normally, I'd hardcode them, but I'd like this component to be aware of its siblings via xstate. Now I have a way to pass state related data into my component, but I'm wondering if there's a sanctioned way to pull sibling states? e.g. I have an `overlay` state that has 

```js
states : {
  hidden : {},
  account : {},
  settings : {},
  help : {},
}
```

I'd essentially like to not have to bake in `[ "ACCOUNT", "SETTINGS", "HELP" ] ` as the navigation options into my nav component and would instead like to have the current state node "inform" what options are available (if I'm on `overlay.account` the nav component can "know" that it's other two options are `settings` and `help`, I'd probably filter off `hidden`). I thought about using `state.can` but this `overlay` state is one of a few parallel nodes, so I can't just brute force filter off of `can` since the number of possible next states takes into account all other parallel states. 

So what I ended up doing is grabbing an interpreted instance I slap onto the window (via `window.state`) (`window.state = interpret(machine)` ) and now I'm reaching into the guts of it with `state.configuration`.

After doing that, I filtered on all my nodes for my overlay node with `const [ node ] = state.configuration.filter(({ id }) => id === "statechart.overlay")`, grabbing the node off of that, and running `Object.Keys` on the `states`. This is incredibly ridiculous but I get that what I'm trying to do is _SUPER CUTE_ so I expect some lawless solutions here. 

Is there a more sanctioned way to do this, or rather, _how dangerous is reaching into the guts of the statechart here?_

**Comment by davidkpiano** [comment, marked as answer]
> This is incredibly ridiculous but I get that what I'm trying to do is SUPER CUTE so I expect some lawless solutions here.
> 
> Is there a more sanctioned way to do this, or rather, how dangerous is reaching into the guts of the statechart here?

Honestly, without any metadata, this is the way to go. The statechart is just a fancy recursive mapped data structure, it _is_ safe to "reach into the guts" and reach for that information in the `StateNode`.

However, I would probably have a thin layer of abstraction in userland:

```js
const machine = createMachine({
  // ...
  states: {
    overlay: createStateWithChildStateMeta({ // bikeshed this
      states: {
        hidden: {},
        account: {},
        settings: {}, 
        // ...
      }
    })
  }
});
```

Where `createStateWithChildStateMeta(stateConfig)` might look like:

```js
function createStateWithChildStateMeta(stateConfig) {
  return {
    meta: { list: Object.keys(stateConfig.states) },
    ...stateConfig
  }
})
```

And then read this from `state.meta`:

```js
const list = getListFromMeta(state); // exercise to reader; it's on state.meta.[stateId].list
// => ['hidden', 'account', 'settings', ...]
```

**Reply by Morklympious** [reply]
Thanks so much for this brain dump, David!
